Support restricted attachments properly in the new Search UI

SammyIsConfused opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should be reported on the issue tracker:
https://github.com/openequella/openEQUELLA/issues

Contributors guide: https://github.com/openequella/openEQUELLA/blob/develop/CONTRIBUTING.md
-->

##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [x] screenshots are included showing significant UI changes

##### Description of change
The new search UI doesn't handle restricted attachments properly - they will still appear in a search, although the file itself will be inaccessible. 
Added to the convertToAttachment function to filter out these attachments when the user does not have VIEW_RESTRICTED_ATTACHMENTS granted.
<!--
Provide a description of the change below this comment. Please include a reference to the GitHub
issue here (not in the title) so as to utilise automatic linking.
-->

<!-- Reference Links -->

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

SammyIsConfused commented:
Need to fix up NewSearchPageTest to account for this (was planning to add some tests for this anyway).

edalex-ian commented:
No probs, will wait until your finish that and request review before I circle back around.

SammyIsConfused commented:
Should be ready now @edalex-ian. The initial problem was that the `attachmentBeans.removeIf(a -> a.isRestricted)` was running even when `attachmentBeans` was null, which broke the test and threw a null pointer. Fixed that, and also added tests to verify that: 
- for a user that doesn't have VIEW_RESTRICTED_ATTACHMENTS the attachment doesn't appear in a search at all
- for a user that does have VIEW_RESTRICTED_ATTACHMENTS the attachment does appear in a search

SammyIsConfused commented:
No more edits - just fixed up the description of the test.

SammyIsConfused commented:
Weird, GHA doesn't want to spin up this time round :thinking:. I'll reword my test description again, see if that wakes it up.

SammyIsConfused commented:
Alright, no idea what happened there, but we have GHA again. Once again ready for review :+1:

SammyIsConfused commented:
Alright, no idea what happened there, but we have GHA again. Once again ready for review :+1:

PenghaiZhang commented on Source/Plugins/Core/com.equella.base/src/com/tle/web/api/item/interfaces/beans/AttachmentBean.java:
```diff
@@ -67,6 +68,14 @@ public void setPreview(boolean preview) {
     this.preview = preview;
   }
 
+  public void setRestricted(boolean restricted) {
```
May I ask where this method gets called ?  I image it would be somewhere in `AttachmentSerializerProvider - serializeAttachment()` .

SammyIsConfused replied:
That's correct, yeah. It iwas initially added to EquellaAttachmentBean, but we use one level up at AttachmentBean for this, so I added it. In the debugger I could already see this value in AttachmentBean, we just didn't have the getter and setter to access it.

PenghaiZhang commented on Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:
```diff
@@ -217,6 +219,12 @@ object SearchHelper {
+    if (CollectionUtils.isNotEmpty(attachmentBeans) && LegacyGuice.aclManager
+          .filterNonGrantedPrivileges(AttachmentConfigConstants.VIEW_RESTRICTED_ATTACHMENTS)
+          .isEmpty) {
+      attachmentBeans.removeIf(a => a.isRestricted)
```
An alternative is changing the type of `attachmentBeans` to List[AttachmentBean], and then we can do
```
val beans = if(no permission) attachmentBeans.filter(a => !a.isRestricted) else attachmentBeans
Option(beans).map(...)
```

SammyIsConfused replied:
As discussed in the office, I moved the filtering inside the Option so that we could get rid of the null check and use it as a Scala list.

PenghaiZhang commented on autotest/OldTests/src/test/java/io/github/openequella/search/NewSearchPageTest.java:
```diff
@@ -88,4 +88,33 @@ public void searchWithLessACLS() {
+  @Test(
+      description =
+          "Search with two differently privileged users when there is a restricted attachment",
+      dependsOnMethods = "searchWithLessACLS")
```
If this test case does not really depend on `searchWithLessACLS`, I think it's better to remove this dependency.

SammyIsConfused replied:
Agreed. Removed the dep.

PenghaiZhang reviewed (commented):
Looks good Sam. Just have a few questions.

SammyIsConfused commented:
Passes locally but fails on CI :thinking: weird. I'll have another look this morning.

SammyIsConfused commented:
Codebuild failed to pull down package.json dependencies which is a bit weird. Not seen this before. Retrying build. 
![image](https://user-images.githubusercontent.com/24543345/111247717-9f244680-865c-11eb-8bec-a211c7b5262c.png)

edalex-ian commented on Source/Plugins/Core/com.equella.core/js/tsrc/components/ItemAttachmentLink.tsx:
```diff
@@ -93,7 +94,12 @@ const ItemAttachmentLink = ({
     // Lightbox viewer not specified, so go with the default of a simple link.
-    <Link href={url} target="_blank" rel="noreferrer">
+    <Link
+      aria-label="Attachment link"
```
This should still be language stringed. Also, can be useful to keep them meaningful if there's lots. To do that here, I suggest you append the `description` when available. And seeing this is duplicated, probably throw it into a constant.

edalex-ian commented on Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:
```diff
@@ -433,7 +433,11 @@ export default function SearchResult({
             >
-              <ListItemText color="primary" primary={description} />
+              <ListItemText
+                title={description}
```
Why's this added?

SammyIsConfused replied:
Left over from a previous xpath query. Sorry about that. Removed.

edalex-ian commented on Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:
```diff
@@ -220,6 +222,12 @@ object SearchHelper {
+            a =>
+              !a.isRestricted || !LegacyGuice.aclManager
+                .filterNonGrantedPrivileges(AttachmentConfigConstants.VIEW_RESTRICTED_ATTACHMENTS)
+                .isEmpty)
```
It seems excessive to call `LegacyGuice.aclManager                .filterNonGrantedPrivileges(AttachmentConfigConstants.VIEW_RESTRICTED_ATTACHMENTS).isEmpty` for each attachment when it's not related _to_ the attachment. How about just throw it in a constant and then reference here?

(Added bonus that it'll then also have a name which is more readable.)

SammyIsConfused replied:
As discussed, moved to using lazy val instead, so this permissions check only gets called if a restricted attachment is part of this item.

edalex-ian commented on autotest/OldTests/src/test/java/io/github/openequella/pages/search/NewSearchPage.java:
```diff
@@ -188,6 +188,41 @@ public void selectOwner(String ownerName) {
+   *     search results.
+   * @return a By for the attachmentText.
+   */
+  private By xPathForAttachmentText(String attachmentText) {
```
Bit of an odd name there, would expect something along the lines of:

```suggestion
  private By attachmentLinkByText(String text) {
```

edalex-ian reviewed (commented):
Looks good @SammyIsConfused , just a couple of queries and suggestions.

edalex-ian reviewed (approved):
Looks good @SammyIsConfused ! Merge on. :+1:
